File: db_model.py
import sqlite3


def create_database(guild):
    try:
        # Creates database for server
        db_path = 'databases/' + str(guild) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute("""
            CREATE TABLE IF NOT EXISTS emojiActivity (
            emoji TEXT,
            person TEXT,
            datetime TEXT 
            )
            """)
        db_conn.commit()
        db_cursor.close()
        print('Database created for', guild)
    except sqlite3.Error as error:
        print("Failed to create sqlite table", error)


def delete_from_db(context, str_emoji):
    """
    Deletes a list of emojis from database.
    """

    try:
        # Deletes record with found emoji, user who posted emoji, and time when emoji was posted.
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute("""
                    DELETE FROM emojiActivity
                    WHERE rowid = 
                    (
                        SELECT rowid
                        FROM emojiActivity
                        WHERE 
                            emoji = ? AND 
                            person = ? AND 
                            datetime = ?
                        LIMIT 1
                    )
                    """, (str_emoji, str(context.author.id), context.created_at.strftime('%Y-%m-%d')))
        print(f"Record has been removed: ({str_emoji}, "
              f"{str(context.author.id)}, "
              f"{context.created_at.strftime('%Y-%m-%d')})")
        db_conn.commit()
        db_cursor.close()
    except sqlite3.Error as error:
        print("Failed to delete record.", error)


def insert_to_db(context, str_emoji):
    """
    Inserts a list of emojis into database.
    """
    try:
        # Inserts new record with found emoji, user who posted emoji, and time when emoji was posted.
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute("""
                INSERT INTO emojiActivity(emoji, person, datetime)
                VALUES(?, ?, ?)
                """, (str_emoji, str(context.author.id), context.created_at.strftime('%Y-%m-%d')))
        db_conn.commit()
        db_cursor.close()
        print(f"Record has been inserted: ({str_emoji}, "
              f"{str(context.author.id)}, "
              f"{context.created_at.strftime('%Y-%m-%d')})")
    except sqlite3.Error as error:
        print("Failed to insert record", error)


def get_leaderboard(context, emoji):
    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
            SELECT 
                person,
                COUNT(emoji) 
            FROM 
                emojiActivity
            WHERE 
                emoji = '{emoji.id}'
                AND person IS NOT '757326308547100712'
            GROUP BY 
                person
            ORDER BY 
                COUNT(emoji) DESC
            LIMIT 10;
        """)
        rows = db_cursor.fetchall()
        db_conn.commit()
        db_cursor.close()
        return rows
    except sqlite3.Error as error:
        print(error)


def get_getcount_member_alltime(context, member):
    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
             SELECT 
                 COUNT(emoji) 
             FROM 
                 emojiActivity
             WHERE 
                 person = '{member}'
             """)
        count = db_cursor.fetchone()
        db_conn.commit()
        db_cursor.close()
        return count
    except sqlite3.Error as error:
        print(error)


def get_getcount_member_monthly(context, member):
    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
                     SELECT 
                         COUNT(emoji) 
                     FROM 
                         emojiActivity
                     WHERE
                         person = '{member}' AND
                         datetime > date('now', '-1 month')
                     """)
        count = db_cursor.fetchone()
        db_conn.commit()
        db_cursor.close()
        return count
    except sqlite3.Error as error:
        print(error)


def get_getcount_member_weekly(context, member):
    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
                     SELECT 
                         COUNT(emoji) 
                     FROM 
                         emojiActivity
                     WHERE
                         person = '{member}' AND
                         datetime > date('now', '-7 day')
                     """)
        count = db_cursor.fetchone()
        db_conn.commit()
        db_cursor.close()
        return count
    except sqlite3.Error as error:
        print(error)


def get_getcount_server_alltime(context):
    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
                     SELECT 
                         COUNT(emoji) 
                     FROM 
                         emojiActivity
                     """)
        count = db_cursor.fetchone()
        db_conn.commit()
        db_cursor.close()
        return count
    except sqlite3.Error as error:
        print(error)


def get_getcount_server_monthly(context):
    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
                             SELECT 
                                 COUNT(emoji) 
                             FROM 
                                 emojiActivity
                             WHERE
                                 datetime > date('now', '-1 month')
                             """)
        count = db_cursor.fetchone()
        db_conn.commit()
        db_cursor.close()
        return count
    except sqlite3.Error as error:
        print(error)


def get_getcount_server_weekly(context):
    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
                             SELECT 
                                 COUNT(emoji) 
                             FROM 
                                 emojiActivity
                             WHERE
                                 datetime > date('now', '-7 day')
                             """)
        count = db_cursor.fetchone()
        db_conn.commit()
        db_cursor.close()
        return count
    except sqlite3.Error as error:
        print(error)


def get_displaystats_member_all(context, member):
    """
    Queries db for specified user input, returns rows of query.
    """

    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
                 SELECT 
                     emoji,
                     COUNT(emoji) 
                 FROM 
                     emojiActivity
                 WHERE 
                     person = '{member}'
                 GROUP BY 
                     emoji
                 ORDER BY COUNT(emoji) DESC
                 """)
        rows = db_cursor.fetchall()
        db_conn.commit()
        db_cursor.close()
        return rows
    except sqlite3.Error as error:
        print(error)


def get_displaystats_member_monthly(context, member):
    """
    Queries db for specified user input, returns rows of query.
    """

    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
                         SELECT 
                             emoji,
                             COUNT(emoji) 
                         FROM 
                             emojiActivity
                         WHERE
                             person = '{member}' AND
                             datetime > date('now', '-1 month')
                         GROUP BY 
                             emoji        
                         ORDER BY COUNT(emoji) DESC
                         """)
        rows = db_cursor.fetchall()
        db_conn.commit()
        db_cursor.close()
        return rows
    except sqlite3.Error as error:
        print(error)


def get_displaystats_member_weekly(context, member):
    """
    Queries db for specified user input, returns rows of query.
    """

    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute(f"""
                         SELECT 
                             emoji,
                             COUNT(emoji) 
                         FROM 
                             emojiActivity
                         WHERE
                             person = '{member}' AND
                             datetime > date('now', '-7 day')
                         GROUP BY 
                             emoji        
                         ORDER BY COUNT(emoji) DESC
                         """)
        rows = db_cursor.fetchall()
        db_conn.commit()
        db_cursor.close()
        return rows
    except sqlite3.Error as error:
        print(error)


def get_displaystats_server_all(context):
    """
    Queries db for specified user input, returns rows of query.
    """

    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute("""
                 SELECT 
                     emoji,
                     COUNT(emoji) 
                 FROM 
                     emojiActivity
                 GROUP BY 
                     emoji        
                 ORDER BY COUNT(emoji) DESC
                 """)
        rows = db_cursor.fetchall()
        db_conn.commit()
        db_cursor.close()
        return rows
    except sqlite3.Error as error:
        print(error)


def get_displaystats_server_monthly(context):
    """
    Queries db for specified user input, returns rows of query.
    """

    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute("""
                         SELECT 
                             emoji,
                             COUNT(emoji) 
                         FROM 
                             emojiActivity
                         WHERE
                             datetime > date('now', '-1 month')
                         GROUP BY 
                             emoji        
                         ORDER BY COUNT(emoji) DESC
                         """)
        rows = db_cursor.fetchall()
        db_conn.commit()
        db_cursor.close()
        return rows
    except sqlite3.Error as error:
        print(error)


def get_displaystats_server_weekly(context):
    """
    Queries db for specified user input, returns rows of query.
    """

    try:
        db_path = 'databases/' + str(context.guild.id) + '.sqlite'
        db_conn = sqlite3.connect(db_path)
        db_cursor = db_conn.cursor()
        db_cursor.execute("""
                         SELECT 
                             emoji,
                             COUNT(emoji) 
                         FROM 
                             emojiActivity
                         WHERE
                             datetime > date('now', '-7 day')
                         GROUP BY 
                             emoji        
                         ORDER BY COUNT(emoji) DESC
                         """)
        rows = db_cursor.fetchall()
        db_conn.commit()
        db_cursor.close()
        return rows
    except sqlite3.Error as error:
        print(error)

File: db_model.js
const Database = require('better-sqlite3')

function createDatabase(guildId) {
    // console.log(`createDatabase(${guildId}) called.`)

    let db = new Database(`./databases/${guildId}.sqlite`);

    const createStatements = [
        `CREATE TABLE IF NOT EXISTS emojiActivity(emoji TEXT, person TEXT, datetime TEXT)`,
        `CREATE TABLE IF NOT EXISTS serverSettings(setting TEXT, flag INTEGER)`,
    ].map(sql => db.prepare(sql))
    for (const createStatement of createStatements) {
        createStatement.run();
    }

    const insertStatement = db.prepare(`INSERT INTO serverSettings (setting, flag) VALUES (@setting, @flag)`)
    const insertSettings = db.transaction((settings) => {
        for (const setting of settings) insertStatement.run(setting)
    })
    insertSettings([
        {setting: 'countmessages', flag: 1},
        {setting: 'countreacts', flag: 1},
        {setting: 'countselfreacts', flag: 1},
    ])

    db.close()
}

function deleteFromDb(guildId, emojiId, personId, dateTime) {
    // console.log(`deleteFromDb(${guildId}, ${emojiId}) called.`)
    let db = new Database(`./databases/${guildId}.sqlite`);
    const statement = db.prepare(`
        DELETE FROM emojiActivity 
        WHERE rowid = 
        (
            SELECT rowid
            FROM emojiActivity
            WHERE
                emoji = @emoji AND
                person = @person AND
                datetime = @datetime
            LIMIT 1
        )
    `)
    statement.run({
        emoji: emojiId,
        person: personId,
        datetime: dateTime
    })
    db.close()

}

function insertToDb(guildId, emojiId, personId, dateTime) {
    // console.log(`insertToDb(${guildId}, ${emojiId}) called.`)
    let db = new Database(`./databases/${guildId}.sqlite`);
    const statement = db.prepare(`
        INSERT INTO emojiActivity (emoji, person, datetime) 
        VALUES (@emoji, @person, @datetime)
    `)
    statement.run({
        emoji: emojiId,
        person: personId,
        datetime: dateTime
    })
    db.close()

}

function getLeaderboard(guildId, emojiId, clientId, dateTime = null) {
    // console.log(`getLeaderboard(${guildId}, ${emojiId}, ${clientId}, ${dateTime}) called.`)
    let db = new Database(`./databases/${guildId}.sqlite`);
    let cat
    if (dateTime) {
        let statement = db.prepare(`
            SELECT 
                person,
                COUNT(emoji) 
            FROM 
                emojiActivity
            WHERE 
                emoji = @emojiId
                AND person IS NOT @clientId
                AND datetime > @dateTime
            GROUP BY 
                person
            ORDER BY 
                COUNT(emoji) DESC
            LIMIT 10;
        `)
        cat = statement.all({
            emojiId: emojiId,
            clientId: clientId,
            dateTime: dateTime,
        })
    } else {
        let statement = db.prepare(`
            SELECT 
                person,
                COUNT(emoji) 
            FROM 
                emojiActivity
            WHERE 
                emoji = @emojiId
                AND person IS NOT @clientId
            GROUP BY 
                person
            ORDER BY 
                COUNT(emoji) DESC
            LIMIT 10;
        `)
        cat = statement.all({
            emojiId: emojiId,
            clientId: clientId
        })
    }
    // console.log(cat)
    db.close()
    return cat

}

function getGetCount(guildId, userId, dateTime) {
    // console.log(`getGetCount(${guildId}, ${userId}, ${dateTime}) called.`)
    let db = new Database(`./databases/${guildId}.sqlite`);
    let count
    if (userId !== null) {
        let statement = db.prepare(`
            SELECT COUNT(emoji) 
            FROM emojiActivity
            WHERE person = @person AND
            datetime > @datetime
        `)
        count = statement.get({person: userId, datetime: dateTime,})
    } else {
        let statement = db.prepare(`
            SELECT COUNT(emoji) 
            FROM emojiActivity
            WHERE datetime > @datetime
        `)
        count = statement.get({datetime: dateTime,})
    }
    db.close()
    return Object.values(count)[0]


}

function getDisplayStats(guildId, dateTime, userId = null) {
    // console.log(`getDisplayStats(${guildId}, ${dateTime}, ${userId}) called.`)

    let db = new Database(`./databases/${guildId}.sqlite`);
    let cat
    if (userId) {
        let statement = db.prepare(`
            SELECT 
                emoji,
                COUNT(emoji) 
            FROM emojiActivity
            WHERE 
                person = @person
                AND datetime > @datetime
            GROUP BY emoji
            ORDER BY COUNT(emoji) DESC
        `)
        cat = statement.all({
            person: userId,
            datetime: dateTime,
        })
    } else {
        let statement = db.prepare(`
            SELECT 
                emoji,
                COUNT(emoji) 
            FROM emojiActivity
            WHERE datetime > @datetime
            GROUP BY emoji
            ORDER BY COUNT(emoji) DESC
        `)
        cat = statement.all({
            datetime: dateTime,
        })
    }
    db.close()
    return cat
}

function getSetting(guildId, setting) {
    let db = new Database(`./databases/${guildId}.sqlite`);
    const statement = db.prepare(`SELECT flag FROM serverSettings WHERE setting = ?`)
    const flag = statement.get(setting).flag
    db.close()
    return flag
}

function setSetting(guildId, setting, flag) {
    let db = new Database(`./databases/${guildId}.sqlite`);
    const statement = db.prepare(`
        UPDATE serverSettings
        SET flag = @flag
        WHERE setting = @setting
    `)
    statement.run({
        setting: setting,
        flag: flag,
    })
    db.close()
}

module.exports = {
    createDatabase,
    deleteFromDb,
    insertToDb,
    getLeaderboard,
    getGetCount,
    getDisplayStats,
    getSetting,
    setSetting
}